mozilla-mobile/firefox-ios
# Bugfix FXIOS-7127 Preserve URL on tab close and restore for web apps

## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7127)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/15837)

## :bulb: Description
When a tab is closed and the Undo button is pressed to restore it from the bottom toast, there's an incorrect redirect of the URL for web apps using pushState (e.g. YouTube). 

For example, if you visit YouTube, tap the top video, and then close the tab and undo closing the tab, the webView should remain on the selected video. However, prior to this bug fix, the restored tab would be back on the home page of YouTube.

Note that closing and restoring the last tab has an incorrect selection state in the tab tray. That’s a separate bug here https://github.com/mozilla-mobile/firefox-ios/issues/20850 (PR https://github.com/mozilla-mobile/firefox-ios/pull/20854)
 
### Root Cause

This is an asynchronicity issue. The webView's session history was not being saved (background Task) in time before the webView was destroyed for the closed tab. So when the tab was restored, the [interactionState](https://developer.apple.com/documentation/webkit/wkwebview/3752236-interactionstate) (or `tabSession` as we call it) was unavailable.

### Testing

This fix should be tested significantly since there are a lot of edge cases. (I didn't test tab syncing in depth, but I did make another bug ticket for a similar issue with syncing behaviour: [GitHub](https://github.com/mozilla-mobile/firefox-ios/issues/20904) [JIRA](https://mozilla-hub.atlassian.net/browse/FXIOS-9447))

**Test Paths:**

1) Tab Refactor Enabled
2) Tab Refactor Disabled

**Test Cases for Each Path:**

1) Close and Undo Close on a single tab
2) Close and Undo Close on the first tab, a middle tab, and a last tab when multiple tabs are open (2, 3+)
2.5) Do this for both the active tab and an inactive tab
3) Close and Undo Close on a tab using the tab button long press submenu
4) Close all tabs at once, and restore all tabs

Note: For all these test cases, the tab in question must be the active tab. Switching tabs and then closing that tab seemed to already save the state as expected. But it wouldn't hurt to test again.

### Testing Link

You can use this link for testing instead of YouTube: https://stuff.overengineer.dev/mozilla/testcases/pushstate-nav/0.html
(courtesy of https://github.com/mozilla-mobile/firefox-ios/issues/15837#issuecomment-2162552585)

## Fix Demo

https://github.com/mozilla-mobile/firefox-ios/assets/173110554/b9c938f3-f533-416b-9275-33ec9d9385e8

## :pencil: Checklist
You have to check all boxes before merging
